### PR TITLE
Fix collections.abc.Callable parsing in ClassVar

### LINKED ISSUE #314: Failed parsing of `ClassVar[Callable[...]]` within dataclass
## Example code

```python
from collections.abc import Callable
from dataclasses import dataclass
from typing import ClassVar

import tyro


@dataclass
class Cmd:
    x: int
    # y: Callable[[int], str]

    FUNC: ClassVar[Callable[[int], str]] = str


cmd = tyro.cli(Cmd)
```

## Expected result

Tyro CLI just works fine (tyro version 0.9.23).

## Actual result

1) Rather cryptic error on `ClassVar` annotation parsing, despite it should never be part of command:

```python
File "tyro/_resolver.py", line 483, in _concretize_type_params
    return origin[new_args]
           ~~~~~~^^^^^^^^^^
  File "<frozen _collections_abc>", line 481, in __new__
TypeError: Expected a list of types, an ellipsis, ParamSpec, or Concatenate. Got <class 'int'>
```

2) And moreover, if `y` instance variable was to be uncommented we'd have the same cryptic error, instead of the one we get when `Callable` is not subscripted : 
> UnsupportedTypeAnnotationError: Unsupported type annotation for field with name `y`, which is resolved to `<class 'collections.abc.Callable'>`. [...]

### PR DESCRIPTION
**Pull request by brentyi**
Fixes #314: Failed parsing of `ClassVar[Callable[...]]` within dataclass

This PR resolves the cryptic TypeError that occurred when using `collections.abc.Callable` in ClassVar annotations. The issue was that collections.abc.Callable args were flattened for processing but not properly unflattened during type reconstruction.

**Changes:**
- Track when collections.abc.Callable args are flattened in _concretize_type_params
- Restore original nested format when reconstructing types
- Add comprehensive test suite for collections.abc.Callable patterns
- Handle edge cases like zero-parameter callables

Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

File: tests/test_collections_abc_callable.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+import tyro
+from tyro import UnsupportedTypeAnnotationError
+
+# Skip tests for Python <= 3.10
+pytestmark = pytest.mark.skipif(
+    sys.version_info <= (3, 10), reason="collections.abc.Callable requires Python > 3.10"
+)
+
+# Import collections.abc.Callable for Python > 3.10
+if sys.version_info > (3, 10):
```
Comment (brentyi) [inline tests/test_collections_abc_callable.py:16]: @claude can you remove this and the condition above? Instead, let's use the conftest.py filter and only run this test for Python >=3.10
